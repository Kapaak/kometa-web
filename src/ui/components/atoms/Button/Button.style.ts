import { css, styled } from 'styled-components';

import { mapColorVariantToColor } from '~/utils/colors';

import { ButtonProps } from './Button.type';

const ButtonInitial = styled.button.withConfig({
  shouldForwardProp: (prop) =>
    !['color', 'size', 'variant', 'customColor', 'fullWidth'].includes(prop),
})<ButtonProps>(({ theme: { typography }, size }) => ({
  ...(size === 'small' ? { ...typography.body6 } : { ...typography.body3 }),
}));

export const Button = styled(ButtonInitial)`
  display: flex;
  justify-content: center;
  align-items: center;
  align-self: center;
  border: none;
  border-radius: 3rem;
  cursor: ${({ disabled }) => (disabled ? 'auto' : 'pointer')};
  font-weight: 500;
  white-space: nowrap;
  font-size: 1.4rem;
  letter-spacing: 0.04rem;
  width: ${({ fullWidth }) => (fullWidth ? '100%' : 'auto')};

  &:hover {
    transition: all 0.2s ease-in-out;
  }

  ${({ size = 'regular' }) => {
    switch (size) {
      case 'small':
        return css`
          padding: 0.5rem 1.2rem;
          min-width: 15rem;
        `;
      case 'regular':
        return css`
          height: 5.1rem;
          min-width: 14rem;
          padding: 1.5rem 3rem;
        `;
    }
  }}

  ${({ color = 'primary', theme: { colors } }) => {
    switch (color) {
      case 'primary':
        return css`
          background-color: ${colors.primary.main};
          &:hover {
            background-color: ${colors.primary.dark};
          }
        `;
      case 'secondary':
        return css`
          background-color: ${colors.secondary.main};
          &:hover {
            background-color: ${colors.secondary.dark};
          }
        `;
      case 'tetriary':
        return css`
          background-color: ${colors.tetriary.main};
          &:hover {
            background-color: ${colors.tetriary.dark};
          }
        `;
      case 'success':
        return css`
          background-color: ${colors.success.main};
          &:hover {
            background-color: ${colors.success.main};
          }
        `;
      case 'error':
        return css`
          background-color: ${colors.error.main};
          &:hover {
            background-color: ${colors.error.main};
          }
        `;
      case 'grey':
        return css`
          color: ${colors.grey[900]};
          background-color: ${colors.grey[200]};
          &:hover {
            background-color: ${colors.grey[200]};
          }
        `;
    }
  }}


${({
    disabled,
    variant = 'filled',
    color = 'primary',
    theme: { colors },
    customColor, // customColor is a prop that can be passed to the component
  }) => {
    const selectedColorVariant = mapColorVariantToColor.call(
      null,
      colors,
      disabled ? 'grey' : color
    );

    switch (variant) {
      case 'plain':
        return css`
          border-radius: 0;
          min-height: 0;
          min-width: 0;
          padding: 0;
          background-color: initial;
          text-transform: none;
          color: ${customColor ?? selectedColorVariant.main};

          &:hover {
            background-color: initial;
          }
        `;
      case 'filled':
        return css`
          color: ${customColor
            ? customColor
            : color === 'grey'
              ? colors.grey[900]
              : colors.grey[100]};
        `;

      case 'outlined':
        return css`
          border: 1px solid ${customColor ?? selectedColorVariant.main};
          color: ${customColor ?? selectedColorVariant.main};
          background-color: initial;
          &:hover {
            background-color: ${disabled
              ? 'initial'
              : selectedColorVariant.main};
            color: ${colors.grey[100]};
          }
        `;
    }
  }}
`;

export const ChildrenContainer = styled.span.withConfig({
  shouldForwardProp: (prop) => !['loading'].includes(prop),
})<{ loading: boolean }>`
  display: inline-flex;
  align-items: center;
  gap: 1rem;
  visibility: ${({ loading }) => (loading ? 'hidden' : 'visible')};
`;

export const FadeLoaderWrapper = styled.div.withConfig({
  shouldForwardProp: (prop) => !['size'].includes(prop),
})<{ size: ButtonProps['size'] }>`
  position: absolute;
  transform: ${({ size }) => (size === 'small' ? 'scale(.4)' : 'scale(.5)')};
`;
