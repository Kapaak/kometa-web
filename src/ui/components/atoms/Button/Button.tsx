import { PropsWithChildren } from 'react';

import { useTheme } from 'styled-components';

import { Loader } from '../Loader';

import * as S from './Button.style';

import { ButtonProps } from './Button.type';

export function Button({
  children,
  loading,
  size = 'regular',
  disabled,
  ...rest
}: PropsWithChildren<ButtonProps>) {
  const theme = useTheme();
  const { grey } = theme.colors;
  return (
    <S.Button size={size} disabled={disabled} {...rest}>
      {loading && (
        <S.FadeLoaderWrapper size={size}>
          <Loader loading={loading} color={grey['100']} />
        </S.FadeLoaderWrapper>
      )}

      <S.ChildrenContainer loading={Boolean(loading)}>
        {children}
      </S.ChildrenContainer>
    </S.Button>
  );
}
