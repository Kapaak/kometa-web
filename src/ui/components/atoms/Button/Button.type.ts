import { ButtonHTMLAttributes } from 'react';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  color?: 'primary' | 'secondary' | 'tetriary' | 'success' | 'error' | 'grey';
  size?: 'small' | 'regular';
  variant?: 'filled' | 'plain' | 'outlined';
  customColor?: string;
  loading?: boolean;
}
