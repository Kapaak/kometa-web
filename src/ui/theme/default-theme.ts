import { DefaultTheme } from 'styled-components';

import { calculateEmFromPx } from '~/utils/dimensions';

export const theme: DefaultTheme = {
  colors: {
    primary: {
      light: '#DCE9FA',
      main: '#014B8B',
      dark: '#1A5FBF',
    },
    secondary: {
      light: '#FFCA0F',
      main: '#FFCA0F',
      dark: '#e6b812',
    },
    tetriary: {
      light: '#73F0C0',
      main: '#73F0C0',
      dark: '#73F0C0',
    },
    error: {
      light: '#FF0000',
      main: '#FF0000',
      dark: '#FF0000',
    },
    success: {
      light: '#34C759',
      main: '#34C759',
      dark: '#34C759',
    },
    grey: {
      100: '#fff',
      200: '#F8F8F8',
      400: '#EDEDED',
      500: '#DADADA',
      700: '#9A9A9A',
      900: '#000000',
    },
  },
  typography: {
    h1: {
      fontSize: '2.4rem',
      fontWeight: 600,
      lineHeight: '3.6rem',
    },
    h2: {
      fontSize: '5.6rem',
      fontWeight: 400,
      lineHeight: '4.8rem',
    },
    h3: {
      fontSize: '3rem',
      fontWeight: 400,
      lineHeight: '4.8rem',
    },
    body1: {
      fontSize: '1.8rem',
      fontWeight: 400,
      lineHeight: '2.4rem',
    },
    body2: {
      fontSize: '1.6rem',
      fontWeight: 400,
      lineHeight: '2.4rem',
    },
    body3: {
      fontSize: '1.6rem',
      fontWeight: 600,
      lineHeight: '3.2rem',
    },
    body4: {
      fontSize: '1.4rem',
      fontWeight: 400,
      lineHeight: '2.4rem',
    },
    body5: {
      fontSize: '1.4rem',
      fontWeight: 500,
      lineHeight: '2.4rem',
    },
    body6: {
      fontSize: '1.4rem',
      fontWeight: 600,
      lineHeight: '2.4rem',
    },
  },
  maxWidth: {
    small: '108rem',
    main: '145rem',
    wide: '145rem',
  },
  shadows: {
    main: '0px 0px 10px 0px rgba(0, 0, 0, 0.05)',
    dark: '0px 0px 15px 0px rgba(0, 0, 0, 0.15)',
  },
  fonts: {
    primary: '"Poppins", sans-serif',
    secondary: '"Bebas Neue", sans-serif',
  },
  breakpoints: {
    sm: calculateEmFromPx(600),
    md: calculateEmFromPx(790),
    lg: calculateEmFromPx(1024),
    xl: calculateEmFromPx(1200),
    xxl: calculateEmFromPx(1920),
  },
};
