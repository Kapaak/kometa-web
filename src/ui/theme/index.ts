export * from './default-theme';
export * from './global';
export * from './style-reset';
export * from './theme-wrapper';
