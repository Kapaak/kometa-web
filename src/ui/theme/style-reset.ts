import { css } from 'styled-components'

export const styleReset = css`
  *,
  *::before,
  *::after {
    margin: 0;
    padding: 0;
    padding: 0;
    border: none;
    font: inherit;
    text-transform: inherit;
    letter-spacing: inherit;
    color: inherit;
    appearance: none;
    border-radius: 0;
    background: none;
  }
`
