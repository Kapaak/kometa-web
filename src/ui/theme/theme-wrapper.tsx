import { ThemeProvider } from 'styled-components'

import { theme } from './default-theme'

export function withTheme(children: React.ReactElement) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}
