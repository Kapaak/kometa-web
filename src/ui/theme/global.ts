import { createGlobalStyle, css } from 'styled-components';

import { styleReset } from './style-reset';

export const GlobalStyle = createGlobalStyle(
  ({ theme: { colors, fonts } }) => css`
    ${styleReset}

    *,
    *::before,
    *::after {
      box-sizing: border-box;
      font-family: ${fonts.primary};
    }

    html {
      font-size: 62.5%;
      min-height: 100%;
      height: 100%;
      scroll-behavior: smooth;
      box-sizing: inherit;
    }

    body {
      min-height: 100%;
      height: 100%;
      font: 400 1.6rem ${fonts.primary};
      text-rendering: optimizeLegibility;
      -webkit-font-smoothing: antialiased;
      -moz-osx-font-smoothing: grayscale;
      font-smooth: always;
      color: ${colors.grey['900']};
      background-color: ${colors.grey['200']};
    }

    //this class puts padding on body and it moves the whole page
    .yarl__no_scroll {
      padding-right: 0 !important;
    }

    br {
      //fixes br height in Firefox to be the same as in Chromium browsers
      margin-bottom: 2.3rem;
    }

    strong {
      font-weight: 600;
    }

    #__next {
      min-height: 100%;
      height: 100%;
      //To prevent the scale down if some item overflows
      /* overflow: auto; */
      //This broke pages and on next page it would retain scroll position
      //TODO: FIX
    }

    a,
    a[href^='tel'],
    a[href^='mail'] {
      color: inherit;
      text-decoration: none;
    }

    input:-webkit-autofill,
    input:-webkit-autofill:hover,
    input:-webkit-autofill:focus,
    input:-webkit-autofill:active {
      -webkit-box-shadow: 0 0 0 3rem ${colors.grey[100]} inset;
    }
  `
);
