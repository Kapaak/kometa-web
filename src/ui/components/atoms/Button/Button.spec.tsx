import { render } from '@testing-library/react'

import { withTheme } from '~/ui/theme'

import { Button } from './Button'

describe('Button', () => {
  it('should render successfully', () => {
    const { baseElement } = render(withTheme(<Button variant="plain" />))

    expect(baseElement).toBeTruthy()
    expect(baseElement).toMatchSnapshot()
  })
})
