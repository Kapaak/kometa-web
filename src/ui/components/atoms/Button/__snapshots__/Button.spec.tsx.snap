// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render successfully 1`] = `
<body>
  <div>
    <button
      class="sc-63139522-0 sc-63139522-1 jAaKDT jYLPWX"
    >
      <span
        class="sc-63139522-2 flvQkd"
      />
    </button>
  </div>
</body>
`;
